File: frontend/src/mocks/refdata.ts
// Mock of the single /refdata endpoint (D-005) plus the user-extensible masters.
// Seed values are verbatim from MASTER-DATA.md — the frontend carries no
// vocabulary of its own; MasterSelect resolves options through here so no
// component ever inlines an option list (CLAUDE.md hard rule / §6).

export interface RefOption {
  value: string;
  label: string;
}

export interface Master {
  id: string;
  /** Human label for the field/master. */
  label: string;
  /** User-extensible masters allow "create new" (institution, sector, tag). */
  extensible: boolean;
  options: RefOption[];
}

// snake_case / kebab value → Title Case label for display.
export function humanize(value: string): string {
  return value
    .replace(/[-_]/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

// Per-vocab value→label overrides — the OFFLINE-fallback mirror of the backend
// `/refdata` `_VOCAB_LABEL_OVERRIDES` (refdata.py). Kept in sync so the fallback
// renders the same served label the live endpoint does (D-005). `will_status:none`
// reads "Not recorded", not the humanized "None" (page-estate §12es-3).
const VALUE_LABEL_OVERRIDES: Record<string, Record<string, string>> = {
  will_status: { none: "Not recorded" },
};

function fixed(values: string[], overrides?: Record<string, string>): RefOption[] {
  return values.map((v) => ({ value: v, label: overrides?.[v] ?? humanize(v) }));
}

// --- Fixed vocabularies (served via /refdata; MASTER-DATA §2) ------------------
const FIXED: Record<string, string[]> = {
  txn_type: [
    "buy", "sell", "dividend", "interest", "deposit", "withdrawal", "fee",
    "split", "bonus", "merger", "transfer",
  ],
  asset_class: [
    "equity", "etf", "mutual_fund", "bond", "cash", "fixed_deposit",
    "commodity", "crypto", "property", "private", "retirement", "liability",
    "other",
  ],
  asset_subclass: ["crypto", "derivative", "equity", "etf", "mutual_fund", "reit"],
  liquidity_profile: ["listed", "redeemable", "locked", "illiquid", "manual"],
  entity_kind: ["self", "spouse", "trust", "company", "other"],
  goal_basis: ["net_worth", "liquid", "none"],
  obligation_recurrence: ["once", "monthly", "quarterly", "annual"],
  obligation_kind: ["expense", "income"],
  contribution_frequency: ["monthly", "quarterly", "annual", "once"],
  contribution_kind: ["invest", "withdraw", "prepay"],
  will_status: ["none", "draft", "executed", "needs_update"],
  estate_doc_status: ["present", "missing", "outdated"],
  valuation_method: [
    "market_quote", "official_nav", "broker_quote", "manual_valuation",
    "statement_import", "calculated_accrual", "estimated_value", "fx_reference",
    "unavailable",
  ],
  entitlement: ["real-time", "delayed", "end-of-day", "cached", "unavailable"],
  policy_dimension: ["asset_class", "currency", "region"],
  cost_basis_method: ["fifo", "average"],
  account_kind: [
    "brokerage", "bank", "retirement", "wallet", "property", "manual", "other",
  ],
  policy_type: [
    "term_life", "whole_life", "health", "critical_illness", "disability",
    "personal_accident", "property", "motor", "travel", "other",
  ],
  premium_frequency: ["monthly", "quarterly", "annual", "single"],
  policy_status: ["active", "lapsed", "expired"],
  estate_doc_category: [
    "will", "insurance", "property", "loan", "identity", "bank", "tax",
    "medical", "other",
  ],
  contact_role: ["nominee", "beneficiary", "executor", "emergency", "guardian"],
  // ND-3 — per-instrument source-override routing (offline fallback; live values
  // from /refdata, sourced from the market-router CAPABILITIES). `auto` clears it.
  source_override: [
    "auto", "alphavantage", "amfi_nav", "coingecko", "csv", "ecb_fx", "eodhd",
    "kite", "mock", "yahoo",
  ],
};

const FIXED_LABELS: Record<string, string> = {
  txn_type: "Transaction type",
  asset_class: "Asset class",
  asset_subclass: "Asset subclass",
  liquidity_profile: "Liquidity profile",
  entity_kind: "Entity kind",
  goal_basis: "Goal basis",
  obligation_recurrence: "Recurrence",
  obligation_kind: "Obligation kind",
  contribution_frequency: "Contribution frequency",
  contribution_kind: "Contribution kind",
  will_status: "Will status",
  estate_doc_status: "Document status",
  valuation_method: "Valuation method",
  entitlement: "Entitlement",
  source_override: "Source override",
  policy_dimension: "Policy dimension",
  cost_basis_method: "Cost-basis method",
  account_kind: "Account kind",
  policy_type: "Policy type",
  premium_frequency: "Premium frequency",
  policy_status: "Policy status",
  estate_doc_category: "Document category",
  contact_role: "Contact role",
};

// --- Currency master (D-006; 22 codes, 9 base-eligible; MASTER-DATA §3) --------
const BASE_ELIGIBLE = ["SGD", "USD", "INR", "EUR", "GBP", "JPY", "AUD", "CNY", "HKD"];
const WIDER = [
  "CAD", "CHF", "AED", "MYR", "THB", "KRW", "TWD", "SEK", "NOK", "DKK", "ZAR",
  "BRL", "NZD",
];
export const CURRENCIES: RefOption[] = [...BASE_ELIGIBLE, ...WIDER].map((c) => ({
  value: c,
  label: c,
}));
export const BASE_CURRENCIES: RefOption[] = BASE_ELIGIBLE.map((c) => ({
  value: c,
  label: c,
}));

// --- Region (derived; six buckets, D-083; MASTER-DATA §4) ----------------------
const REGIONS = ["India", "Singapore", "US", "Europe", "APAC", "Other"];

// --- Extensible masters (MASTER-DATA §6) ---------------------------------------
// Sector: 11 GICS sectors (authored, PROPOSED, D-009/DEF-6). Institution/Tag
// start empty in real data; seeded here so the picker has something to show.
const SECTORS = [
  "Energy", "Materials", "Industrials", "Consumer Discretionary",
  "Consumer Staples", "Health Care", "Financials", "Information Technology",
  "Communication Services", "Utilities", "Real Estate",
];
const INSTITUTIONS = [
  "DBS Bank", "OCBC", "Interactive Brokers", "Zerodha", "Vanguard",
  "Standard Chartered",
];

export const MASTERS: Record<string, Master> = {
  ...Object.fromEntries(
    Object.entries(FIXED).map(([id, values]) => [
      id,
      { id, label: FIXED_LABELS[id] ?? humanize(id), extensible: false, options: fixed(values, VALUE_LABEL_OVERRIDES[id]) },
    ]),
  ),
  currency: { id: "currency", label: "Currency", extensible: false, options: CURRENCIES },
  base_currency: { id: "base_currency", label: "Base currency", extensible: false, options: BASE_CURRENCIES },
  region: { id: "region", label: "Region", extensible: false, options: REGIONS.map((r) => ({ value: r, label: r })) },
  sector: { id: "sector", label: "Sector", extensible: true, options: SECTORS.map((s) => ({ value: s, label: s })) },
  institution: { id: "institution", label: "Institution", extensible: true, options: INSTITUTIONS.map((i) => ({ value: i, label: i })) },
};

export function getMaster(id: string): Master {
  const m = MASTERS[id];
  if (!m) throw new Error(`Unknown master: ${id}`);
  return m;
}

File: app/api/v1/routes/refdata.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Reference data (D-005): the single endpoint serving every fixed vocabulary.

Retires the per-master `*/meta` endpoints and all frontend inline lists — the
frontend carries zero vocabulary copies (MASTER-DATA §1). Values are sourced from
the canonical code enums/constants where they exist; the authored vocabularies
(asset_subclass DEF-2, and the sets with no single code home) are defined here
against MASTER-DATA, which is their backend home.
"""

from __future__ import annotations

from fastapi import APIRouter

from app.core.config import SUPPORTED_CURRENCIES
from app.models import AssetClass, TxnType
from app.schemas.common import EntitlementStatus, ValuationMethod
from app.services.accounts import ACCOUNT_KINDS, COST_BASIS_METHODS
from app.services.entities import ENTITY_KINDS
from app.services.estate import (
    CONTACT_ROLES,
    DOC_CATEGORIES,
    DOC_STATUSES,
    WILL_STATUSES,
)
from app.services.insurance import FREQUENCIES as PREMIUM_FREQUENCIES
from app.services.insurance import POLICY_STATUSES, POLICY_TYPES
from app.services.planning import GOAL_BASES, OBLIGATION_KINDS, RECURRENCES

router = APIRouter()


def _values(enum_cls) -> list[str]:
    return [m.value for m in enum_cls]


# Item 3b (2026-07-10) — /refdata serves DISPLAY LABELS alongside values so the UI
# never hardcodes a mapping (D-005). Most labels titleize the snake_case value;
# acronyms/brands that titleize wrong are overridden here (the single source).
_LABEL_OVERRIDES = {
    "etf": "ETF", "reit": "REIT", "amfi_nav": "AMFI", "coingecko": "CoinGecko",
    "alphavantage": "AlphaVantage", "ecb_fx": "ECB FX", "eodhd": "EODHD", "csv": "CSV",
    "isin": "ISIN", "cusip": "CUSIP", "figi": "FIGI", "sedol": "SEDOL",
    "amfi_code": "AMFI code", "kite_token": "Kite token", "coingecko_id": "CoinGecko ID",
    "provider_symbol": "Provider symbol", "us": "US", "us-europe": "US / Europe",
}


def _label(value: str) -> str:
    ov = _LABEL_OVERRIDES.get(value.lower())
    if ov:
        return ov
    # Titleize snake_case / all-lower enums (mutual_fund -> "Mutual fund"); leave
    # already display-ready values (India, Singapore, APAC, SGD) untouched.
    if "_" in value or value.islower():
        s = value.replace("_", " ")
        return s[:1].upper() + s[1:]
    return value


# Per-vocab label overrides (MASTER-DATA §2) — a value whose SERVED display label
# differs from the titleized default IN THE CONTEXT of one vocabulary. Kept per-vocab
# (not global like `_LABEL_OVERRIDES`) so an override can never leak into a sibling
# vocab that happens to share the value. `will_status:none` reads "Not recorded"
# (the honesty framing), not the bare "None" — ratified at the Estate specimen
# geometry walk (page-estate §12es-3, 2026-07-16); the UI renders this served label
# verbatim and never re-maps it (D-005).
_VOCAB_LABEL_OVERRIDES: dict[str, dict[str, str]] = {
    "will_status": {"none": "Not recorded"},
    # page-accounts §9-13: the titleizer serves "Fifo"; the acronym reads "FIFO" (the UI renders
    # the served label verbatim, §12es-3). `average` titleizes correctly and is left alone.
    "cost_basis_method": {"fifo": "FIFO"},
}


def _labeled(values: list[str], overrides: dict[str, str] | None = None) -> list[dict[str, str]]:
    ov = overrides or {}
    return [{"value": v, "label": ov.get(v) or _label(v)} for v in values]


def label_for(vocab: str, value: str) -> str:
    """The SERVED display label for one value of a fixed vocabulary — the SAME truth `/refdata`
    serves through (the per-vocab overrides + the titleizer), exposed for backend consumers that
    compose UI server-side. The Reports Pack composer resolves attribution asset-class keys through
    this so a rendered label can never drift from the one the JSON routes serve (reports-pack §12pk-1
    / the §12es-3 label-truth rule; the csv_import → `_TXN_APPLICABILITY` cross-import precedent)."""
    ov = _VOCAB_LABEL_OVERRIDES.get(vocab, {})
    return ov.get(value) or _label(value)


# Authored / no-single-code-home vocabularies (MASTER-DATA §2/§4).
_ASSET_SUBCLASS = ["crypto", "derivative", "equity", "etf", "mutual_fund", "reit"]  # DEF-2 §2 †
_LIQUIDITY_PROFILE = ["listed", "redeemable", "locked", "illiquid", "manual"]  # §2
_ENTITY_KIND = ENTITY_KINDS  # §2 — single source (Amendment H, the policy_status pattern)
_CONTRIBUTION_FREQUENCY = ["monthly", "quarterly", "annual", "once"]  # §2
_CONTRIBUTION_KIND = ["invest", "withdraw", "prepay"]  # §2
_COST_BASIS_METHOD = COST_BASIS_METHODS  # §2 (v2 lanes; `spec` is ROADMAP R-6) — single source
_POLICY_DIMENSION = ["asset_class", "currency", "region"]  # §2
_REGION = ["India", "Singapore", "US", "Europe", "APAC", "Other"]  # §4 (D-083)
_ID_TYPE = ["isin", "cusip", "figi", "sedol", "amfi_code", "kite_token",
            "coingecko_id", "provider_symbol"]  # §2


def _source_override_values() -> list[str]:
    """ND-3 — per-instrument source-override routing options. `auto` clears the
    override; the rest are the market-router providers (sourced from CAPABILITIES so
    the vocab can never drift from the router)."""
    from app.providers.market.router import CAPABILITIES
    return ["auto", *sorted(CAPABILITIES.keys())]

# D-090 (ratified 2026-07-10) — AssetClass → the TxnTypes the Add-flow Type
# dropdown OFFERS for that class. MASTER-DATA §10 is the canonical statement; this
# is its backend home so the frontend carries no copy (D-005). Form-level filtering
# ONLY — the engine (`compute_fifo`) is unchanged and processes every type, and CSV
# imports of odd-but-real historical events are NOT filtered by this matrix
# (importer validates and commits regardless). Ratification amendment: ETF Bonus is
# ON (ETF unit splits/consolidations occur). Crypto corporate actions OFF (enable on
# request); retirement/liability interest ON; MF split+bonus ON; FD/bond/cash
# interest ON (the Manual-branch transaction path this implies is approved).
_TXN_APPLICABILITY: dict[str, list[str]] = {
    "equity":        ["buy", "sell", "dividend", "fee", "split", "bonus", "merger", "transfer"],
    "etf":           ["buy", "sell", "dividend", "fee", "split", "bonus", "merger", "transfer"],
    "mutual_fund":   ["buy", "sell", "dividend", "fee", "split", "bonus", "merger", "transfer"],
    "bond":          ["buy", "sell", "interest", "fee", "transfer"],
    "cash":          ["interest", "deposit", "withdrawal", "fee", "transfer"],
    "fixed_deposit": ["interest", "deposit", "withdrawal", "fee", "transfer"],
    "commodity":     ["buy", "sell", "fee", "transfer"],
    "crypto":        ["buy", "sell", "fee", "transfer"],
    "property":      ["buy", "sell", "fee", "transfer"],
    "private":       ["buy", "sell", "fee", "transfer"],
    "retirement":    ["interest", "deposit", "withdrawal", "fee", "transfer"],
    "liability":     ["interest", "deposit", "withdrawal", "fee", "transfer"],
    "other":         ["buy", "sell", "dividend", "interest", "deposit", "withdrawal",
                      "fee", "split", "bonus", "merger", "transfer"],
}


@router.get("/refdata")
async def refdata() -> dict[str, list[dict[str, str]]]:
    """Every fixed vocabulary, keyed by id, each as `{value, label}` pairs so the UI
    renders served DISPLAY LABELS and never hardcodes a mapping (D-005; item 3b).
    Extensible masters (institution, sector, tag) are served by their own endpoints, not here.

    **`currency` IS served here** (MASTER-DATA §3 AMENDMENT 2026-07-14): the currency master is the
    fixed code constant ``SUPPORTED_CURRENCIES`` — the reference TABLE the spec used to describe was
    never built — so currency is a **fixed vocabulary** and belongs on `/refdata` like any other
    (D-005: the frontend carries zero vocabulary copies). It is what a policy `currency` bucket is
    validated against (Gate A9), so the picker and the validator now read the SAME list."""
    raw = {
        "txn_type": _values(TxnType),
        "asset_class": _values(AssetClass),
        "asset_subclass": _ASSET_SUBCLASS,
        "liquidity_profile": _LIQUIDITY_PROFILE,
        "entity_kind": _ENTITY_KIND,
        "goal_basis": list(GOAL_BASES),
        "obligation_recurrence": list(RECURRENCES),
        "obligation_kind": list(OBLIGATION_KINDS),
        "contribution_frequency": _CONTRIBUTION_FREQUENCY,
        "contribution_kind": _CONTRIBUTION_KIND,
        "will_status": list(WILL_STATUSES),
        "estate_doc_status": list(DOC_STATUSES),
        "estate_doc_category": list(DOC_CATEGORIES),
        "contact_role": list(CONTACT_ROLES),
        "valuation_method": _values(ValuationMethod),
        "entitlement": _values(EntitlementStatus),
        "policy_dimension": _POLICY_DIMENSION,
        "currency": list(SUPPORTED_CURRENCIES),
        "region": _REGION,
        "cost_basis_method": _COST_BASIS_METHOD,
        "account_kind": list(ACCOUNT_KINDS),
        "policy_type": list(POLICY_TYPES),
        "premium_frequency": list(PREMIUM_FREQUENCIES),
        "policy_status": list(POLICY_STATUSES),
        "id_type": _ID_TYPE,
        "source_override": _source_override_values(),
    }
    return {vocab: _labeled(values, _VOCAB_LABEL_OVERRIDES.get(vocab)) for vocab, values in raw.items()}


@router.get("/refdata/txn-applicability")
async def txn_applicability() -> dict[str, list[str]]:
    """D-090 — per-AssetClass TxnType applicability for the Add-flow Type dropdown.
    Keyed by asset_class value → the offered txn_type values (MASTER-DATA §10).
    Form-level filtering only; the engine and the CSV importer ignore it."""
    return _TXN_APPLICABILITY
